Userland/Libraries/LibGL/SoftwareGLContext.cpp

```diff
@@ -3009,4 +3006,173 @@ void SoftwareGLContext::sync_device_texcoord_config()
+        break;
+    case GL_LINEAR_ATTENUATION:
+        device_light.linear_attenuation = param;
+        break;
+    case GL_QUADRATIC_ATTENUATION:
+        device_light.quadratic_attenuation = param;
+        break;
+    case GL_SPOT_EXPONENT:
+        device_light.spotlight_exponent = param;
+        break;
+    case GL_SPOT_CUTOFF:
+        device_light.spotlight_cutoff_angle = param;
+        break;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+void SoftwareGLContext::gl_lightfv(GLenum light, GLenum pname, const GLfloat* params)
+{
+    RETURN_WITH_ERROR_IF(m_in_draw_state, GL_INVALID_OPERATION);
+    RETURN_WITH_ERROR_IF(light > GL_LIGHT7, GL_INVALID_ENUM);
+    RETURN_WITH_ERROR_IF(!(pname == GL_AMBIENT || pname == GL_DIFFUSE || pname == GL_SPECULAR || pname == GL_POSITION || pname == GL_CONSTANT_ATTENUATION || pname == GL_LINEAR_ATTENUATION || pname == GL_QUADRATIC_ATTENUATION || pname == GL_SPOT_CUTOFF || pname == GL_SPOT_EXPONENT || pname == GL_SPOT_DIRECTION), GL_INVALID_ENUM);
+
+    auto& device_light = m_rasterizer.light(light - GL_LIGHT0);
+
+    switch (pname) {
+    case GL_AMBIENT:
+        device_light.ambient_intensity = { params[0], params[1], params[2], params[3] };
+        break;
+    case GL_DIFFUSE:
+        device_light.diffuse_intensity = { params[0], params[1], params[2], params[3] };
+        break;
+    case GL_SPECULAR:
+        device_light.specular_intensity = { params[0], params[1], params[2], params[3] };
+        break;
+    case GL_POSITION:
+        device_light.position = { params[0], params[1], params[2], params[3] };
+        device_light.position = m_model_view_matrix * device_light.position;
+        break;
+    case GL_CONSTANT_ATTENUATION:
+        device_light.constant_attenuation = *params;
+        break;
+    case GL_LINEAR_ATTENUATION:
+        device_light.linear_attenuation = *params;
+        break;
+    case GL_QUADRATIC_ATTENUATION:
+        device_light.quadratic_attenuation = *params;
+        break;
+    case GL_SPOT_EXPONENT:
+        device_light.spotlight_exponent = *params;
+        break;
+    case GL_SPOT_CUTOFF:
+        device_light.spotlight_cutoff_angle = *params;
+        break;
+    case GL_SPOT_DIRECTION: {
+        FloatVector4 direction_vector = { params[0], params[1], params[2], 0.0f };
+        direction_vector = m_model_view_matrix * direction_vector;
+        device_light.spotlight_direction = { direction_vector.x(), direction_vector.y(), direction_vector.z() };
+        break;
+    }
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+void SoftwareGLContext::gl_materialf(GLenum face, GLenum pname, GLfloat param)
+{
+    RETURN_WITH_ERROR_IF(!(face == GL_FRONT || face == GL_BACK || face == GL_FRONT_AND_BACK), GL_INVALID_ENUM);
+    RETURN_WITH_ERROR_IF(pname != GL_SHININESS, GL_INVALID_ENUM);
+    RETURN_WITH_ERROR_IF(param > 128.0f, GL_INVALID_VALUE);
+
+    switch (face) {
+    case GL_FRONT:
+        m_rasterizer.material(0).shininess = param;
```

Comment:
We should probably have a descriptive enum instead of `0` and `1` for `GL_FRONT` and `GL_BACK`